apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionTests.AbortedTransactionRetryTests.cs

```diff
@@ -0,0 +1,337 @@
+                    Assert.Contains("Bang!", exception.Message);
+                });
+
+                callee.AssertCallTimesInrange(new Tuple<DateTime, DateTime>(time0, time0));
+                Assert.Equal(time0, scheduler.Clock.GetCurrentDateTimeUtc());
+            }
+
+            private SpannerConnection BuildSpannerConnection(Mock<SpannerClient> spannerClientMock)
+            {
+                var spannerClient = spannerClientMock.Object;
+                var sessionPoolOptions = new SessionPoolOptions
+                {
+                    MaintenanceLoopDelay = TimeSpan.Zero
+                };
+
+                var sessionPoolManager = new SessionPoolManager(sessionPoolOptions, spannerClient.Settings.Logger, (_o, _s, _l) => Task.FromResult(spannerClient));
+                sessionPoolManager.SpannerSettings.Scheduler = spannerClient.Settings.Scheduler;
+                sessionPoolManager.SpannerSettings.Clock = spannerClient.Settings.Clock;
+
+                SpannerConnectionStringBuilder builder = new SpannerConnectionStringBuilder
+                {
+                    DataSource = DatabaseName.Format(SpannerClientHelpers.ProjectId, SpannerClientHelpers.Instance, SpannerClientHelpers.Database),
+                    SessionPoolManager = sessionPoolManager
+                };
+
+                return new SpannerConnection(builder);
+            }
+
+            private class Callee
+            {
+
+                private readonly FakeScheduler _scheduler;
+                private readonly IList<DateTime> _callTimes = new List<DateTime>();
+
+                public Callee(FakeScheduler scheduler) => _scheduler = scheduler;
+
+                public async Task<int> DatabaseWorkAsync(SpannerConnection connection, SpannerTransaction transaction)
+                {
+                    _callTimes.Add(_scheduler.Clock.GetCurrentDateTimeUtc());
+
+                    var insert = connection.CreateInsertCommand("table_1",
+                        new SpannerParameterCollection
+                        {
+                        new SpannerParameter("column_1", SpannerDbType.Int64, 10)
+                        });
+                    insert.Transaction = transaction;
+
+                    var dml = transaction.CreateBatchDmlCommand();
+                    dml.Add("UPDATE table_1 SET column_1 = column1 + 5");
+
+                    await insert.ExecuteNonQueryAsync();
+                    await dml.ExecuteNonQueryAsync();
+
+                    return _callTimes.Count;
+                }
+
+                public Task<int> Fails(SpannerTransaction transaction)
+                {
+                    _callTimes.Add(_scheduler.Clock.GetCurrentDateTimeUtc());
+                    throw new InvalidOperationException("Bang!");
+                }
+
+                public void AssertCallTimesInrange(params Tuple<DateTime, DateTime>[] times)
```

Comment:
Nit: capitalize the R of Range - but we may change how all of this works anyway :)